src/content/1.4/code/agda/full.agda

```diff
@@ -1,27 +1,32 @@
 open import Data.Char using (Char; toUpper)
 open import Data.String using (String; _++_; fromList; toList; words)
-open import Data.Product using (_×_; _,_)
+open import Data.Product using (_×_; _,_) renaming (proj₁ to fst; proj₂ to snd)
 open import Data.List using (List) renaming (map to lmap)
-open import Function using (_∘_)
+open import Function using (_∘_; _$_)
 
-variable a b : Set
+private variable a b c : Set
 
 Writer : Set → Set
-Writer a = ( a × String )
+Writer a = a × String
 
-morphism : ∀ (a b : Set) → a → Writer b
-morphism = {!   !}
+morphism : a → Writer b
+morphism = {!!}
 
-_>=>_ : ∀ {a b c : Set} → (a → Writer b) → (b → Writer c) → (a → Writer c)
+_>=>_ : (a → Writer b) → (b → Writer c) → (a → Writer c)
 m1 >=> m2 = λ x →
   let (y , s1) = m1 x
       (z , s2) = m2 y
   in (z , s1 ++ s2)
 
+-- alternative --
+_≫=≫_ : (a → Writer b) → (b → Writer c) → a → Writer c
+m1 ≫=≫ m2 = λ x →
+  fst (m2 $ fst (m1 x)) , snd (m1 x) ++ snd (m2 $ fst (m1 x))
```

Comment:
We are computing `m1 x` three times here.

I think version with `where` is more idiomatic to Agda. At least it is often used in: https://www.cs.bham.ac.uk/~mhe/HoTT-UF-in-Agda-Lecture-Notes/

Reply:
Okay, I can remove that one.  I just thought it was worth seeing another way to present it.

As for computing `m1 x` three times, imho, when it comes to formalizing mathematics, computational efficiency should play no role... unless we are formalizing mathematics of computational efficiency! :)